quixstreams/sources/community/file/s3.py

```diff
@@ -0,0 +1,171 @@
+            roughly reflects the time "delay" between the original message producing.
+            Use any float >= 0, where 0 is no delay, and 1 is the original speed.
+            NOTE: Time delay will only be accurate per partition, NOT overall.
+        :param name: The name of the Source application (Default: last folder name).
+        :param shutdown_timeout: Time in seconds the application waits for the source
+            to gracefully shut down.
+        :param on_client_connect_success: An optional callback made after successful
+            client authentication, primarily for additional logging.
+        :param on_client_connect_failure: An optional callback made after failed
+            client authentication (which should raise an Exception).
+            Callback should accept the raised Exception as an argument.
+            Callback must resolve (or propagate/re-raise) the Exception.
+        """
+        super().__init__(
+            filepath=filepath,
+            key_setter=key_setter,
+            value_setter=value_setter,
+            timestamp_setter=timestamp_setter,
+            file_format=file_format,
+            compression=compression,
+            has_partition_folders=has_partition_folders,
+            replay_speed=replay_speed,
+            name=name or f"s3_{bucket}_{Path(filepath).name}",
+            shutdown_timeout=shutdown_timeout,
+            on_client_connect_success=on_client_connect_success,
+            on_client_connect_failure=on_client_connect_failure,
+        )
+        self._bucket = bucket
+        self._credentials = {
+            "region_name": region_name,
+            "aws_access_key_id": aws_access_key_id,
+            "aws_secret_access_key": aws_secret_access_key,
+            "endpoint_url": endpoint_url,
+        }
+        self._client: Optional[S3Client] = None
+
+    def _close_client(self):
+        logger.debug("Closing S3 client session...")
+        if self._client:
+            try:
+                self._client.close()
+            except Exception as e:
+                logger.error(f"Azure client session exited non-gracefully: {e}")
+        self._client = None
+
+    def setup(self):
+        if self._client is None:
+            self._client = boto_client("s3", **self._credentials)
+            try:
+                self._client.head_bucket(Bucket=self._bucket)
+            except botocore.exceptions.ClientError as e:
+                if "404" in str(e):
+                    raise MissingS3Bucket(f'S3 bucket "{self._bucket}" does not exist.')
+                raise
+
+    def get_file_list(self, filepath: Union[str, Path]) -> Iterable[Path]:
+        resp = self._client.list_objects(
+            Bucket=self._bucket,
+            Prefix=str(filepath),
+            Delimiter="/",
+        )
+        for _folder in resp.get("CommonPrefixes", []):
+            yield from self.get_file_list(_folder["Prefix"])
```

Comment:
why do we have an underscore here?
```suggestion
        for folder in resp.get("CommonPrefixes", []):
            yield from self.get_file_list(folder["Prefix"])
```

Reply:
leftover from some previous name overlaps, whoops